CMakeLists.txt

```diff
@@ -1,84 +1,58 @@
 # SPDX-FileCopyrightText: 2008-2021 HPDCS Group <rootsim@googlegroups.com>
 # SPDX-License-Identifier: GPL-3.0-only
 cmake_minimum_required(VERSION 3.10)
-project("ROOT-Sim core" LANGUAGES C)
-set(PROJECT_VERSION 3.0.0-alpha.3)
-set(PROJECT_DESCRIPTION "A General-Purpose Multithreaded Parallel/Distributed Simulation Library")
+project("ROOT-Sim core" LANGUAGES C DESCRIPTION "A General-Purpose Multi-threaded Parallel/Distributed Simulation Library")
 
-#set(CMAKE_VERBOSE_MAKEFILE ON)
-
-if (MSVC)
-    # warning level 4 and all warnings as errors
-    add_compile_options(/Wall)
-else()
-    # lots of warnings and all warnings as errors
-    add_compile_options(-Wall -Wextra -pedantic)
-endif()
-
-add_definitions(-DROOTSIM_VERSION="${PROJECT_VERSION}")
-
-set(srcs
-        src/arch/io.c
-        src/arch/mem.c
-        src/arch/thread.c
-        src/core/core.c
-        src/init.c
-        src/core/sync.c
-        src/datatypes/msg_queue.c
-        src/distributed/control_msg.c
-        src/distributed/mpi.c
-        src/gvt/fossil.c
-        src/gvt/gvt.c
-        src/gvt/termination.c
-        src/lib/lib.c
-        src/lib/random/random.c
-        src/lib/state/state.c
-        src/lib/topology/topology.c
-        src/log/file.c
-        src/log/log.c
-        src/log/stats.c
-        src/lp/lp.c
-        src/lp/process.c
-        src/mm/auto_ckpt.c
-        src/mm/buddy/buddy.c
-        src/mm/buddy/ckpt.c
-        src/mm/msg_allocator.c
-        src/parallel/parallel.c
-        src/serial/serial.c)
+set(PROJECT_VERSION 3.0.0-alpha.3) # todo: use cmake project VERSION argument when we are out of the alpha
 
 set(CMAKE_C_STANDARD 11)
 set(CMAKE_C_STANDARD_REQUIRED ON)
 set(CMAKE_C_VISIBILITY_PRESET hidden)
-#set(CMAKE_INTERPROCEDURAL_OPTIMIZATION ON)
-#set(CMAKE_UNITY_BUILD ON)
-#set(CMAKE_UNITY_BUILD_BATCH_SIZE 128)
-
-find_package(Threads REQUIRED)
-find_package(MPI REQUIRED)
+# set(CMAKE_UNITY_BUILD ON)
+# set(CMAKE_UNITY_BUILD_BATCH_SIZE 128)
+
+include(CheckIPOSupported)
+check_ipo_supported(RESULT SUPPORTS_IPO OUTPUT ERROR_IPO)
+if(SUPPORTS_IPO)
+    set(CMAKE_INTERPROCEDURAL_OPTIMIZATION_RELEASE ON)
+    set(CMAKE_C_COMPILE_OPTIONS_IPO "-flto")
+else()
+    message(WARNING "LTO is not supported")
+endif()
 
 if(MSVC)
     message(FATAL_ERROR "You are using the MSVC compiler, which does not properly implement C11. Please consider switching to clang.")
 endif()
 
-add_compile_options(${TARGET_NAME} ${MPI_C_COMPILE_FLAGS})
-link_libraries(${MPI_C_LIBRARIES} ${CMAKE_THREAD_LIBS_INIT})
-if(NOT WIN32)
-    link_libraries(m)
+find_package(Threads REQUIRED)
+find_package(MPI REQUIRED)
+
+add_compile_options(-Wall -Wextra -pedantic)
+
+set(LOG_LEVEL_DEFINE
+    $<$<CONFIG:Debug>:LOG_LEVEL=LOG_TRACE>
+    $<$<CONFIG:RelWithDebInfo>:LOG_LEVEL=LOG_TRACE>
+    $<$<CONFIG:Release>:LOG_LEVEL=LOG_INFO>
+    $<$<CONFIG:MinSizeRel>:LOG_LEVEL=LOG_INFO>
+)
```

Comment:
Maybe I am missing something: I remember that we switched to `global_config.log_level` to manage the log level. Is this macro actually used?

Reply:
I just checked and you're right, that macro isn't used anymore. I will remove this code in the next commit.

As a side note: we used to have both a ```log_level``` global variable and a ```LOG_LEVEL_DEFINE``` macro so that, in Release mode, verbose log messages could be compiled away. Probably, the minimal performance gains are not worth the loss in clarity anyway.

Reply:
I agree, I don't think it's worth it. So we can remove this part!